### DISCUSSION TITLE
How to grab all classes defined from command line?

### DISCUSSION THREAD

**Discussion by craigcomstock** [cfengine/core#4520] [Q&A] [answered]
I want to write a bundle which responds to any classes defined in a `cf-agent` command line.

I know if I do `cf-promises --show-classes -Dfoo | grep environment` I get:

```
craig@libre:~$ cf-promises --show-classes -Dfoo | grep environment
foo                                                          source=environment    
```

And there is the [classesmatching](https://docs.cfengine.com/docs/3.15/reference-functions-classesmatching.html) function that can filter on tags and [classmatch](https://docs.cfengine.com/docs/3.15/reference-functions-classmatch.html) which mentions `hardclass` but is there some way to grab all the environment defined classes?

**Comment by olehermanse** [comment, marked as answer]
You want them in an slist, right?

You can use the tag in classesmatching (including the `=` part):

```
$ cat test.cf
bundle agent __main__
{
  vars:
    "cmd_line_classes"
      slist => sort(classesmatching(".*", "source=environment"));
  reports:
    "The classes: $(with)"
      with => join(", ", cmd_line_classes);
}
$ /var/cfengine/bin/cf-agent -KI test.cf -Dfoo,bar,baz
R: The classes: bar, baz, foo
```

(I don't think it's possible to preserve the ordering from the command line, but I guess that doesn't matter).

**Reply by craigcomstock** [reply]
Yep. From the docs I didn't get any clues that "source=environment" is a "tag" by itself.

Thanks!
